### ISSUE TITLE
Regression: special case of `or` pattern no longer typechecks

### ISSUE THREAD

**Issue by philnguyen** [racket/typed-racket#150]
The following program typechecks in Racket snapshot `6.2.0.3--2015-05-22(78ecccc/a)`, but no longer does in `6.2.0.4--2015-06-10(bf12a2b/a)`:

```{racket}
#lang typed/racket

(: f : (Listof Integer) (Listof Integer) → Integer)
(define (f xs ys)
  (match* (xs ys)
    [((list a b) (or (list a b) (list b a))) (+ a b)]
    [(_ _) 42]))
```

The error message looks like that from the known problem with the `or` pattern in general:
```
; /tmp/test.rkt:6:48: Type Checker: type mismatch
;   expected: Integer
;   given: (U Integer False)
;   in: a
; /tmp/test.rkt:6:50: Type Checker: type mismatch
;   expected: Integer
;   given: (U Integer False)
;   in: b
; Type Checker: Summary: 2 errors encountered
```

**Comment by samth** [comment]
This is almost certainly because of the recent change to the expansion of `match` -- maybe I need to fix the order of `or`.

**Comment by takikawa** [comment]
Looks like this hasn't been resolved yet. Do we need to fix this for 6.3?

**Comment by samth** [comment]
Yes, I should fix this.
